README
NAME
    Test::Mock::Class - Simulating other classes

SYNOPSIS
      use Test::Mock::Class ':all';
      mock_class 'Net::FTP' => 'Net::FTP::Mock';
      my $mock_object = Net::FTP::Mock->new;

      # anonymous mocked class
      my $metamock = mock_anon_class 'Net::FTP';
      my $mock_object = $metamock->new_object;

      # anonymous class with role applied
      my $metamock = Test::Mock::Class->create_anon_class(
          roles => [ 'My::Handler::Role' ],
      );
      my $mock_object = $metamock->new_object;

DESCRIPTION
    In a unit test, mock objects can simulate the behavior of complex, real
    (non-mock) objects and are therefore useful when a real object is
    impractical or impossible to incorporate into a unit test.

    The unique features of `Test::Mock::Class':

    * Its API is inspired by PHP SimpleTest framework.

    * It isn't tied with Test::Builder so it can be used standalone or with
      any xUnit-like framework, i.e. Test::Unit::Lite. Look for
      Test::Builder::Mock::Class if you want to use it with Test::Builder
      (Test::More or Test::Simple).

    * The API for creating mock classes is based on Moose and Class::MOP so
      it doesn't clash with API of original class and is easy expandable.

    * The methods for defining mock object's behavior are prefixed with
      `mock_' string so they shouldn't clash with original object's methods.

    * Mocks as actors: The mock version of a class has all the methods of
      the original class. The return value will be `undef', but it can be
      changed with `mock_returns' method.

    * Mocks as critics: The method of mock version of a class can check its
      calling arguments and throws an exception if arguments don't match
      (`mock_expect' method). An exception also can be thrown if the method
      wasn't called at all (`mock_expect_once' method).

BUGS
    The API is not stable yet and can be changed in future.

AUTHOR
    Piotr Roszatycki <dexter@cpan.org>

LICENSE
    Based on SimpleTest, an open source unit test framework for the PHP
    programming language, created by Marcus Baker, Jason Sweat, Travis
    Swicegood, Perrick Penet and Edward Z. Yang.

    Copyright (c) 2009 Piotr Roszatycki <dexter@cpan.org>.

    This program is free software; you can redistribute it and/or modify it
    under GNU Lesser General Public License.



Changes
0.04 | Piotr Roszatycki <dexter@cpan.org> | 2011-01-25

Changes:

* Works with Any::Moose.
* Module namespace::functions replaced namespace::clean.

------------------------------------------------------------------------
0.0303 | Piotr Roszatycki <dexter@cpan.org> | 2010-05-01

Fixes:

* Class::MOP requires IO::* modules explicitly.  Implicit load does not work
  correctly.

------------------------------------------------------------------------
0.0302 | Piotr Roszatycki <dexter@cpan.org> | 2009-11-17

Fixes:

* Requires Class::MOP => 0.93, Moose >= 0.90.

------------------------------------------------------------------------
0.0301 | Piotr Roszatycki <dexter@cpan.org> | 2009-10-27

Fixes:

* Compatibility with Moose >= 0.81.

Changes:

* Test::Mock::Class::Role::Meta::Class:
  - Method "_mock_reinitialize" is not necessary anymore.

------------------------------------------------------------------------
0.03 | Piotr Roszatycki <dexter@cpan.org> | 2009-04-21

Fixes:

* Build requires Class::Inspector.

Changes:

* Requires Exception::Base and Exception::Fatal.

* Count all matched expectations:

  $m->mock_expect_once('method', args => [1, 2, 3]);
  $m->mock_expect_call_count('method', 2);
  $m->method(1, 2, 3);  # matches first expectation
  $m->method();
  $m->mock_tally();     # matches second expectation 

* Test::Assert's assertions are used to check if arguments matches.      

------------------------------------------------------------------------
0.02 | Piotr Roszatycki <dexter@cpan.org> | 2009-04-18

Changes:

* Method "create_mock_anon_class" accepts "role" argument.
* Function "mock_anon_class" has optional argument.
* Mock class reapplies roles of original class.
* Mock class has original class and Moose::Object as superclasses.
* Class::Inspector is not required anymore.

------------------------------------------------------------------------
0.0102 | Piotr Roszatycki <dexter@cpan.org> | 2009-04-16

Fixes:

* Requires Class::Inspector.
* This module is under LGPL only.

------------------------------------------------------------------------
0.0101 | Piotr Roszatycki <dexter@cpan.org> | 2009-04-15

Fixes:

* Requires namespace::clean and Test::Assert 0.05.

------------------------------------------------------------------------
0.01 | Piotr Roszatycki <dexter@cpan.org> | 2009-04-14

* First CPAN release.

------------------------------------------------------------------------


t/test_moose.pl
#!/usr/bin/perl

use 5.006;

use strict;
use warnings;

use File::Basename;
use File::Spec;
use Cwd;

BEGIN {
    chdir dirname(__FILE__) or die "$!";
    chdir '..' or die "$!";

    unshift @INC, map { /(.*)/; $1 } split(/:/, $ENV{PERL5LIB}) if defined $ENV{PERL5LIB} and ${^TAINT};

    my $cwd = ${^TAINT} ? do { local $_=getcwd; /(.*)/; $1 } : '.';
    unshift @INC, File::Spec->catdir($cwd, 'inc');
    unshift @INC, File::Spec->catdir($cwd, 'lib');
}

use Test::Unit::Lite;

$ENV{ANY_MOOSE} = 'Moose';

local $SIG{__WARN__} = sub { require Carp; Carp::confess(@_) };

all_tests;


t/test_mouse.pl
#!/usr/bin/perl

use 5.006;

use strict;
use warnings;

use File::Basename;
use File::Spec;
use Cwd;

BEGIN {
    chdir dirname(__FILE__) or die "$!";
    chdir '..' or die "$!";

    unshift @INC, map { /(.*)/; $1 } split(/:/, $ENV{PERL5LIB}) if defined $ENV{PERL5LIB} and ${^TAINT};

    my $cwd = ${^TAINT} ? do { local $_=getcwd; /(.*)/; $1 } : '.';
    unshift @INC, File::Spec->catdir($cwd, 'inc');
    unshift @INC, File::Spec->catdir($cwd, 'lib');
}

use Test::Unit::Lite;

$ENV{ANY_MOOSE} = 'Mouse';

local $SIG{__WARN__} = sub { require Carp; Carp::confess(@_) };

all_tests;


lib/Test/Mock/Class.pm
#!/usr/bin/perl -c

package Test::Mock::Class;

=head1 NAME

Test::Mock::Class - Simulating other classes

=head1 SYNOPSIS

  use Test::Mock::Class ':all';
  require Net::FTP;

  # concrete mocked class
  mock_class 'Net::FTP' => 'Net::FTP::Mock';
  my $mock_object = Net::FTP::Mock->new;

  # anonymous mocked class
  my $metamock = mock_anon_class 'Net::FTP';
  my $mock_object = $metamock->new_object;

  # anonymous class with role applied
  my $metamock = Test::Mock::Class->create_anon_class(
      roles => [ 'My::Handler::Role' ],
  );
  my $mock_object = $metamock->new_object;

=head1 DESCRIPTION

In a unit test, mock objects can simulate the behavior of complex, real
(non-mock) objects and are therefore useful when a real object is impractical
or impossible to incorporate into a unit test.

The unique features of C<Test::Mock::Class>:

=over 2

=item *

Its API is inspired by PHP SimpleTest framework.

=item *

It isn't tied with L<Test::Builder> so it can be used standalone or with any
xUnit-like framework, i.e. L<Test::Unit::Lite>.  Look for
L<Test::Builder::Mock::Class> if you want to use it with L<Test::More> or
other L<Test::Builder>-based libraries.

=item *

The API for creating mock classes is based on L<Moose> and L<Class::MOP> so it
doesn't clash with API of original class and is easy expandable.
C<Test::Mock::Class> works also with L<Mouse> and then it doesn't require
compiled XS modules.

=item *

The methods for defining mock object's behavior are prefixed with C<mock_>
string so they shouldn't clash with original object's methods.

=item *

Mocks as actors: The mock version of a class has all the methods of the
original class.  The return value will be C<undef>, but it can be changed with
C<mock_returns> method.

=item *

Mocks as critics: The method of mock version of a class can check its calling
arguments and throws an exception if arguments don't match (C<mock_expect>
method).  An exception also can be thrown if the method wasn't called at all
(C<mock_expect_once> method).

=back

=for readme stop

=cut

use 5.006;

use strict;
use warnings;

our $VERSION = '0.04';

use Any::Moose;


=head1 INHERITANCE

=over 2

=item *

extends L<Moose::Meta::Class>

=cut

extends any_moose('::Meta::Class');

=item *

with L<Test::Mock::Class::Role::Meta::Class>

=back

=cut

with 'Test::Mock::Class::Role::Meta::Class';


use namespace::functions -except => 'meta';


=head1 FUNCTIONS

=over

=cut

BEGIN {
    my %exports = ();

=item B<mock_class>( I<class> : Str, I<mock_class> : Str = undef ) : Moose::Meta::Class

Creates the concrete mock class based on original I<class>.  If the name of
I<mock_class> is undefined, its name is created based on name of original
I<class> with added C<::Mock> suffix.

The original I<class> is loaded with C<L<Class::MOP>::load_class> function
which behaves wrongly for some packages, i.e. I<IO::File>.  It is much safer
to require original class explicitly.

The function returns the metaclass object of new I<mock_class>.

=cut

    $exports{mock_class} = sub {
        sub ($;$) {
            return __PACKAGE__->create_mock_class(
                defined $_[1] ? $_[1] : $_[0] . '::Mock',
                class => $_[0],
            );
        };
    };

=item B<mock_anon_class>( I<class> : Str = undef ) : Moose::Meta::Class

Creates an anonymous mock class based on original I<class>.  The name of this
class is automatically generated.  If I<class> argument not defined, the empty
mock class is created.

The function returns the metaobject of new mock class.

=back

=cut

    $exports{mock_anon_class} = sub {
        sub (;$) {
            return __PACKAGE__->create_mock_anon_class(
                defined $_[0] ? (class => $_[0]) : (),
            );
        };
    };

=head1 IMPORTS

=over

=cut

    my %groups = ();

=item Test::Mock::Class ':all';

Imports all functions into caller's namespace.

=back

=cut

    $groups{all} = [ keys %exports ];

    require Sub::Exporter;
    Sub::Exporter->import(
        -setup => {
            exports => [ %exports ],
            groups => \%groups,
        },
    );
};


no namespace::functions;

1;


=begin umlwiki

= Class Diagram =

[                                 Test::Mock::Class
 ----------------------------------------------------------------------------------
 ----------------------------------------------------------------------------------
 <<utility>> mock_class(class : Str, mock_class : Str = undef) : Moose::Meta::Class
 <<utility>> mock_anon_class(class : Str) : Moose::Meta::Class
                                                                                   ]

[Test::Mock::Class] ---|> [Moose::Meta::Class] [<<role>> Test::Mock::Class::Role::Meta::Class]

=end umlwiki

=head1 EXAMPLE

The C<Test::Mock::Class> fits perfectly to L<Test::Unit::Lite> tests.  It
throws an exception immediately if some problem is occurred.  It means that
the test unit is failed if i.e. the mock method is called with wrong
arguments.

Example code:

  package My::ExampleTest;

  use Test::Unit::Lite;

  use Moose;
  extends 'Test::Unit::TestCase';

  use Test::Assert ':all';
  use Test::Mock::Class ':all';

  require IO::File;

  sub test_mock_class {
      my ($self) = @_;

      my $mock = mock_anon_class 'IO::File';
      my $io = $mock->new_object;
      $io->mock_return( open => 1, args => [qr//, 'r'] );

      assert_true( $io->open('/etc/passwd', 'r') );

      $io->mock_tally;
  };

=head1 SEE ALSO

Mock classes for L<Test::More>: L<Test::Builder::Mock::Class>.

Mock metaclass API: L<Test::Mock::Class::Role::Meta::Class>,
L<Moose::Meta::Class>.

Mock object methods: L<Test::Mock::Class::Role::Object>.

xUnit-like testing: L<Test::Unit::Lite>.

Other implementations: L<Test::MockObject>, L<Test::MockClass>.

=for readme continue

=head1 BUGS

The API is not stable yet and can be changed in future.

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Based on SimpleTest, an open source unit test framework for the PHP
programming language, created by Marcus Baker, Jason Sweat, Travis Swicegood,
Perrick Penet and Edward Z. Yang.

Copyright (c) 2009, 2010, 2011 Piotr Roszatycki <dexter@cpan.org>.

This program is free software; you can redistribute it and/or modify it
under GNU Lesser General Public License.


Build.PL
#!/usr/bin/perl

use 5.006;

use strict;
use warnings;

use Module::Build;

my $module = 'Test::Mock::Class';

(my $dist = $module) =~ s/::/-/g;

my $build = Module::Build->new(
    module_name => $module,
    license => 'lgpl',
    requires => {
        'perl'                 => 5.006002,
        'namespace::functions' => 0,
        'Any::Moose'           => 0,
        'Exception::Base'      => 0,
        'Exception::Fatal'     => 0,
        'Test::Assert'         => 0.05,
    },
    optional => {
    },
    build_requires => {
        'Class::Inspector'     => 0,
        'Test::Unit::Lite'     => 0.12,
    },
    auto_configure_requires => 0,
    create_makefile_pl => 'traditional',
    create_readme => 1,
    create_license => 1,
    sign => 1,
    recursive_test_files => 1,
    add_to_cleanup => [
        '*.bak', '*.gz', '*.tgz', '*.ppd', '*.ppmx', '*.deb', '*.bat',
        'Makefile', 'pm_to_blib', 'cover_db', 'xt/Debian_CPANTS.txt',
        "$dist-*",
    ],
    meta_merge => {
        resources => {
            bugtracker => "http://rt.cpan.org/NoAuth/Bugs.html?Dist=$dist",
            repository => "http://github.com/dex4er/perl-$dist",
        },
    },
);

$build->create_build_script;


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.35
require 5.006;
use ExtUtils::MakeMaker;
WriteMakefile
(
          'TESTS' => 't/*.t t/tlib/*.t t/tlib/Test/*.t t/tlib/Test/Mock/*.t t/tlib/Test/Mock/Class/*.t t/tlib/Test/Mock/Class/Test/*.t t/tlib/Test/Mock/Class/Role/*.t t/tlib/Test/Mock/Class/Role/Object/*.t t/tlib/Test/Mock/Class/Role/Meta/*.t t/tlib/Test/Mock/Class/Role/Meta/Class/*.t',
          'PL_FILES' => {},
          'INSTALLDIRS' => 'site',
          'NAME' => 'Test::Mock::Class',
          'EXE_FILES' => [],
          'VERSION_FROM' => 'lib/Test/Mock/Class.pm',
          'PREREQ_PM' => {
                           'Class::MOP' => '0.93',
                           'Exception::Fatal' => 0,
                           'namespace::clean' => 0,
                           'Test::Unit::Lite' => '0.12',
                           'Class::Inspector' => 0,
                           'Test::Assert' => '0.05',
                           'Exception::Base' => 0,
                           'Moose' => '0.9'
                         }
        )
;
